paramnb/widgets.py

```diff
@@ -0,0 +1,140 @@
+        self._lists[True].options = selected
+        self._lists[True].value = []
+        self._lists[False].options = [o for o in self.options
+                                      if o not in selected]
+
+    def _update_options(self, event):
+        """
+        Updates the options of each of the sublists after the options
+        for the whole widget are updated.
+        """
+        self._reverse_lookup = {v: k for k, v in event['new'].items()}
+        options = event['new'].keys() if isinstance(event, dict) else event['new']
+        self._selected[False] = []
+        self._selected[True] = []
+        self._lists[True].options = ['']
+        self._lists[True].value = []
+        self._lists[False].options = options
+        self._lists[False].value = []
+        self._filter_options()
+
+    def _filter_options(self, event=None):
+        """
+        Filters unselected options based on a text query event.
+        """
+        query = self._query if event is None else event['new'] 
+        self._query = query
+        whitelist = self._lists[True].options
+        options = [o for o in self.options
+                   if o not in whitelist]
+        if not query:
+            self._lists[False].options = options
+            self._lists[False].value = []
+        else:
+            matches = fnmatch.filter(options, query+'*')
+            self._lists[False].options = matches if matches else ['']
+            self._lists[False].value = matches
+
+    def _update_selection(self, event):
+        """
+        Updates the current selection in each list.
+        """
+        selected = event['owner'] is self._lists[True]
+        self._selected[selected] = [v for v in event['new'] if v != '']
+
+    def _apply_selection(self, event):
+        """
+        Applies the current selection depending on which button was
+        pressed.
+        """
+        selected = event is self._buttons[True]
+        new = self._selected[not selected]
+        old = self._lists[selected].options
+        other = self._lists[not selected].options
+        
+        merged = sorted([v for v in list(old) + list(new) if v != ''])
+        leftovers = sorted([o for o in other if o not in new and o != ''])
+        new_values = merged if selected else leftovers
+        self._lists[selected].options = merged if merged else ['']
+        self._lists[not selected].options = leftovers if leftovers else ['']
+        self.value = [self._options_dict[o] for o in self._lists[True].options if o != '']
+        self._filter_options()
+
+    def _ipython_display_(self, **kwargs):
+        """
+        Displays the composite widget.
+        """
+        self._composite._ipython_display_(**kwargs)
+    
+    def get_state(self, key=None, drop_defaults=False):
+        """
+        HACK: Let's this composite widget pretend to be a regular widget
+        when included into a layout.
+        """
+        if key == 'value':
```

Comment:
Anything saying HACK should be a comment, not a docstring; no need to advertise that except to someone looking at the code (and therefore empowered to come up with something better).

Let's -> Lets.

I can't follow the logic of it, but if it works...